service/frontend/workflow_handler.go

```diff
@@ -4988,3 +4926,157 @@ func (wh *WorkflowHandler) unaliasCreateScheduleRequestSearchAttributes(request
+	// it will either signal or start a new workflow with the same ID and parameters as the completed one.
+	// The loop protects against a race condition where the state of workflow changes in between checking the state
+	// of workflow and sending the signal or starting a new workflow.
+	// TODO: However, there is still a rare race condition left which cannot be fixed without having a notion of atomic
+	// workflow start. That is if the state changes from completed(describe stage)->running->completed(start workflow),
+	// where the workflow will be started with stale info acquired from the describe stage.
+	for {
+		executionInfo, startScheduleArgs, err := wh.getScheduleExecutionInfoAndArgs(ctx, workflowID, namespaceName, namespaceID)
+		if err != nil {
+			return err
+		}
+
+		inputPayloads, err := getStartInputPayloads(startScheduleArgs)
+		if err != nil {
+			return err
+		}
+		err = wh.checkEventBlobSizeLimit(ctx, workflowID, namespaceName, namespaceID, signalInputPayloads.Size(), operationName)
+		if err != nil {
+			return err
+		}
+
+		if executionInfo.GetStatus() == enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING {
+			_, err = wh.historyClient.SignalWorkflowExecution(ctx, &historyservice.SignalWorkflowExecutionRequest{
+				NamespaceId: namespaceID.String(),
+				SignalRequest: &workflowservice.SignalWorkflowExecutionRequest{
+					Namespace:         namespaceName,
+					WorkflowExecution: &commonpb.WorkflowExecution{WorkflowId: workflowID, RunId: executionInfo.Execution.RunId},
+					SignalName:        signalName,
+					Input:             signalInputPayloads,
+					Identity:          identity,
+					RequestId:         requestId,
+				},
+			})
+			var notFoundErr *serviceerror.NotFound
+			if err == nil {
+				break
+			}
+			if !errors.As(err, &notFoundErr) && err.Error() == consts.ErrWorkflowCompleted.Error() {
+				return err
+			}
+
+		} else if executionInfo.GetStatus() == enumspb.WORKFLOW_EXECUTION_STATUS_COMPLETED {
+			_, err = wh.historyClient.StartWorkflowExecution(ctx, &historyservice.StartWorkflowExecutionRequest{
+				NamespaceId: namespaceID.String(),
+				StartRequest: &workflowservice.StartWorkflowExecutionRequest{
+					Namespace:             namespaceName,
+					WorkflowId:            workflowID,
+					WorkflowType:          &commonpb.WorkflowType{Name: scheduler.WorkflowType},
+					TaskQueue:             &taskqueuepb.TaskQueue{Name: primitives.PerNSWorkerTaskQueue},
+					Input:                 inputPayloads,
+					Identity:              identity,
+					RequestId:             requestId,
+					SearchAttributes:      executionInfo.SearchAttributes,
+					Memo:                  executionInfo.Memo,
+					WorkflowIdReusePolicy: enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE,
+				},
+			})
+			if err == nil {
+				// if successfully started a new schedule workflow attempt to delete the old workflow and ignore errors
+				wh.historyClient.DeleteWorkflowExecution(ctx, &historyservice.DeleteWorkflowExecutionRequest{
```

Comment:
what if we had the workflow delete the previous execution itself (with a background activity)? then we get retries and some assurance it'll eventually happen (except for bugs)